src/lib/api-keys.ts

```diff
@@ -1,6 +1,74 @@
+import type { Request } from 'express';
 import * as _ from 'lodash';
+import { errors } from '@balena/pinejs';
+import {
+	createApiKey,
+	ApiKeyOptions,
+	PartialCreateKey,
+} from '../platform/api-keys';
+
+const { BadRequestError } = errors;
+
+const supportedActorTypes = ['application', 'user', 'device'] as const;
+
+export type ApiKeyParameters = {
+	actorType: typeof supportedActorTypes[number];
+	actorTypeId: number;
+	roles: string[];
+} & Pick<ApiKeyOptions, 'name' | 'description' | 'apiKey'>;
+
+export const createGenericApiKey = async (
+	req: Request,
+	{
+		actorType,
+		actorTypeId,
+		roles,
+		name,
+		description,
+		apiKey,
+	}: ApiKeyParameters,
+) => {
+	if (!supportedActorTypes.includes(actorType)) {
+		throw new BadRequestError('Unsupported actor type');
+	}
+
+	if (!Number.isFinite(actorTypeId)) {
+		throw new BadRequestError('Actor must be a number');
+	}
+
+	if (!Array.isArray(roles) || roles.some((r) => typeof r !== 'string' || !r)) {
+		throw new BadRequestError('Roles should be an array of non-empty strings');
```

Comment:
This is probably more user friendly if written as
```suggestion
		throw new BadRequestError('Roles should be an array of role names');
```
It provides the same info since a role name is a string and is also non-empty whilst also providing the context of what the string represents

Reply:
Loved it :heart: